Mathlib/Computability/Halting.lean

```diff
@@ -57,6 +57,59 @@ theorem merge' {f g} (hf : Nat.Partrec f) (hg : Nat.Partrec g) :
     · exact ⟨x, by simp only [e, Option.mem_def, Option.orElse_eq_orElse, Option.orElse_none]⟩
     · exact ⟨y, by simp only [Option.orElse_eq_orElse, Option.orElse_some, Option.mem_def]⟩
 
+/-- If a partial recursive function on pairs depends only on the first component,
+it can be represented as a partial recursive function omitting the second argument. -/
+lemma projection {f : ℕ →. ℕ} (hf : Nat.Partrec f)
+    (unif : ∀ {m n₁ n₂ a₁ a₂ : ℕ}, a₁ ∈ f (m.pair n₁) → a₂ ∈ f (m.pair n₂) → a₁ = a₂) :
+    ∃ g : ℕ →. ℕ, Nat.Partrec g ∧ ∀ a m, (a ∈ g m ↔ ∃ z, a ∈ f (m.pair z)) := by
+  obtain ⟨cf, rfl⟩ := Code.exists_code.1 hf
+  let F : ℕ → ℕ → Option ℕ := fun m n ↦
+    Nat.rec .none (fun x ih ↦ ih.casesOn (cf.evaln n (m.pair x)) .some) n
+  have : Primrec₂ F := .to₂
+    <| Primrec.nat_rec' Primrec.snd (.const Option.none)
+      (Primrec.option_casesOn (Primrec.snd.comp .snd)
+        (Code.primrec_evaln.comp
+          <| _root_.Primrec.pair
+            (_root_.Primrec.pair (Primrec.snd.comp .fst) (.const cf))
+            (Primrec₂.natPair.comp (Primrec.fst.comp .fst) (Primrec.fst.comp .snd)))
+        (Primrec.option_some.comp .snd).to₂).to₂
```

Comment:
Do you think you could get this down to
```suggestion
  have : Primrec₂ F := by fun_prop
```
by tagging the relevant predicates and lemmas with `fun_prop`?

Reply:
Aha! Then yes consider this as a great reason to do it now :)